hibernate-rx-core/src/main/java/org/hibernate/rx/engine/impl/RxTwoPhaseLoad.java

```diff
@@ -0,0 +1,543 @@
+/*
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later.
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.rx.engine.impl;
+
+import java.io.Serializable;
+import java.util.Optional;
+import java.util.concurrent.CompletionStage;
+
+import org.hibernate.AssertionFailure;
+import org.hibernate.CacheMode;
+import org.hibernate.HibernateException;
+import org.hibernate.LockMode;
+import org.hibernate.bytecode.enhance.spi.LazyPropertyInitializer;
+import org.hibernate.cache.spi.access.EntityDataAccess;
+import org.hibernate.cache.spi.entry.CacheEntry;
+import org.hibernate.engine.internal.Versioning;
+import org.hibernate.engine.profile.Fetch;
+import org.hibernate.engine.profile.FetchProfile;
+import org.hibernate.engine.spi.EntityEntry;
+import org.hibernate.engine.spi.EntityKey;
+import org.hibernate.engine.spi.LoadQueryInfluencers;
+import org.hibernate.engine.spi.PersistenceContext;
+import org.hibernate.engine.spi.SessionEventListenerManager;
+import org.hibernate.engine.spi.SessionFactoryImplementor;
+import org.hibernate.engine.spi.SharedSessionContractImplementor;
+import org.hibernate.engine.spi.Status;
+import org.hibernate.event.service.spi.EventListenerGroup;
+import org.hibernate.event.service.spi.EventListenerRegistry;
+import org.hibernate.event.spi.EventType;
+import org.hibernate.event.spi.PostLoadEvent;
+import org.hibernate.event.spi.PostLoadEventListener;
+import org.hibernate.event.spi.PreLoadEvent;
+import org.hibernate.event.spi.PreLoadEventListener;
+import org.hibernate.graph.spi.AttributeNodeImplementor;
+import org.hibernate.graph.spi.GraphImplementor;
+import org.hibernate.internal.CoreMessageLogger;
+import org.hibernate.persister.entity.EntityPersister;
+import org.hibernate.pretty.MessageHelper;
+import org.hibernate.property.access.internal.PropertyAccessStrategyBackRefImpl;
+import org.hibernate.proxy.HibernateProxy;
+import org.hibernate.rx.persister.entity.impl.RxEntityPersister;
+import org.hibernate.rx.util.impl.RxUtil;
+import org.hibernate.stat.internal.StatsHelper;
+import org.hibernate.stat.spi.StatisticsImplementor;
+import org.hibernate.type.Type;
+import org.hibernate.type.TypeHelper;
+
+import org.jboss.logging.Logger;
+
+/**
+ * @see org.hibernate.engine.internal.TwoPhaseLoad
+ */
+public class RxTwoPhaseLoad {
```

Comment:
@DavideD I *hate hate hate* the idea of copy/pasting this very complex functionality into Hibernate RX. Is it necessary? If so, we need documentation explaining why it's necessary, and exactly what it does differently to `TwoPhaseLoad`.

I recently put a lot of work into reducing code duplication and adding JavaDoc to things in the Hibernate RX codebase, and this is precisely the sort of code duplication that I was trying to eliminate and/or properly document.